### ISSUE TITLE
Dimmer always 100% from Home.app Automation

### ISSUE THREAD

**Issue by drewgould** [kuestess/homebridge-platform-insteonlocal#47]
First, thanks for the work you've put into this great plugin. I previously had the HomeKit Insteon Hub and got sick of its unreliability and constantly showing No Response. I've got everything setup now with a 2245 hub. 

When I manually turn on a dimmer type (Model # 2477D) switch to a set % in the Home app, it turns on to that %. I have an Automation set up in the Home app to turn my living room lights on to 35% at sunset. When that scene runs, the lights turn on but they always turn on to 100% instead of 35%. 

If the lights are on and I test run the scene, sometimes they'll dim down to the %, sometimes they stay at 100%. Do I have my devices misconfigured or is there something else going on?

{
"name": "Fireplace Lights",
"deviceID": "25.96.B6",
"dimmable": "yes",
"deviceType":"dimmer",
"controllers": []
}

**Comment by drewgould** [comment]
An update here as I had some time to dig into this further. When looking at the logs I noticed an on command was being sent after the dimmer level command:

> 
> [12/26/2018, 5:30:37 PM] [Insteon Local Platform] Setting level of Great Room Lights to 35%
> [12/26/2018, 5:30:37 PM] [Insteon Local Platform] Setting power state of Great Room Lights to on
> 

I believe the issue is line 1912: 
`self.service.getCharacteristic(Characteristic.On).on('set', self.setPowerState.bind(self)).`

It does not appear to be necessary to turn the dimmer on before setting the level. I commented this line out and my dimmers now turn on and are set to the correct level, instead of 100% as I previously reported. As I'm new to this codebase, I didn't want to submit a pull request with a casual understanding of all the Insteon ins and outs. I'm happy to do so if you concur with the change.

**Comment by kuestess** [comment]
@drewgould The Home app loves to send a 'on' command along with the level command which is why you get the 100% on followed by _theoretically_ the set level.  As you know, Insteon will handle the level command on its own and turn the device on if necessary, so the extra 'on' is redundant.  I filter those out in `setPowerState`, but the extra Insteon traffic may prevent the level command from being processed in a reasonable time.  With your change above, does the dimmer still turn on if you tap it in the home app?

**Comment by rasod** [comment]
@drewgould I had the same problem a while back, and made a change similar to yours. Basically ignore on commands from HomeKit for dimmable lights, since the setBrightness command will do the same. However in my case if you ask Siri to turn a light on, only the on command is sent with no following setBrightness.

**Comment by drewgould** [comment]
Yeah this issue is more complicated than I thought (what else is new). @rasod is correct that while it solves the the scene from dimming and then running up to 100%, it does break just tapping a switch to turn it on or off. You have to dim and slide to 100% or 0% to turn on or off. 

@kuestess I do see Homekit sending both the on and brightness 'set' commands. It seems both commands fire off from their event handlers in parallel to the hub and the last one in is whether you get x% or 100%. Since it seems Homekit sends both, perhaps we could debounce the on command for the dimmer and let a brightness command intercept and cancel an 'on' command if it comes in with x milliseconds? Otherwise if it was solely an on command, the command would wait the x ms and then run as normal.

**Comment by kuestess** [comment]
@drewgould Working on exactly that and hope to have something here in the next few days.  I've also changed the 'on' set handler to call the brightness function which should make it easier to intercept and cancel the 100% brightness command assuming that it is quickly followed by a level command.

**Comment by kuestess** [comment]
@drewgould @rasod Just published 0.3.15 to npm that filters out duplicate on commands and adds a debounce for dimming.  Eliminates insteon traffic and seems to be more responsive.  I played around with the debounce timer a little and it seems to work ok - let me know what you think.  Scene performance seems to be much improved.  cc: @shereefalkady

**Comment by drewgould** [comment]
Quick 5 minute reaction is very positive! Automation scene dims properly and my scenes definitely seem to run a bit quicker turning 23 accessories on and off. Will monitor it for a couple of real world on/off cycles but looks good so far

**Comment by shereefalkady** [comment]
@kuestess sounds amazing will try it — how or where can I post a video or image for you to see — currently working on 160 switch house

**Comment by kuestess** [comment]
@shereefalkady That's a lot of switches!  You should be able to attach a pic/video directly to your comment here.

**Comment by kuestess** [comment]
@shereefalkady That's a serious setup!  You're probably best off with an Insteon group for large sets of devices as its much more efficient, but you might test the latest plugin update too and see how it performs.

**Comment by shereefalkady** [comment]
@kuestess sure Insteon scenes for massive groups ... but something like “living room to 33%” living room being 9 switches ...

**Comment by RoccoDiabolo** [comment]
Hi everyone, I'm a total newbie to Homebridge. A friend got me a Raspi for my birthday and helped me set up Homebridge on it and this Insteon plug-in. Everything seemed fine but now I am experiencing a weird problem similar to what was described in this thread. When using the HomeKit app on my iPhone to control a dimmer switch, everything works fine. I can click on it and the light turns on 100%. When using voice commands via Siri instead of the HomeKit app, I get the confirmation that the light is on, however, it is actually not! When looking at the HomeKit app, the icon for that particular light is 'turned on' but at 0%. I tried this with multiple dimmer switches and dimmer outlets with the same result. On/off switches continue to work via Siri.

Could this be a result of the 0.3.15 update? I could have sworn it was working as expected on Dec-29/30 or so when "Hey Siri, turn on Office light" would actually turn on that light. Like I said, I'm totally new to this and it might be operator error but since this thread seems related, I though I would ask.

**Comment by kuestess** [comment]
@RoccoDiabolo So fun times, the Home app loves to send an on AND dim command when you turn on a device, while Siri only sends the on OR dim (which is correct).  Previously, I was just discarding the on as Insteon will handle the dim just fine.  I've reworked the `setBrightness` command to process the 'on' if its the only command sent or filter it out if it gets a dim command within 300ms.  Just published on npm - see if that works for you.

**Comment by RoccoDiabolo** [comment]
@kuestess  - This worked! Thank you so much for the fix! This plug-in is awesome!

**Comment by kuestess** [comment]
@drewgould Still working for you with the latest update?

**Comment by drewgould** [comment]
Been buried in dev work this week but will upgrade tonight and report back!

**Comment by drewgould** [comment]
This is working well for me. I think we can consider this issue resolved!

**Comment by bradgessler** [comment]
I still experience this issue as of rev 1fedbc1dffe6e9959248314fb3d43ce01fde36a9 with the Insteon 2413U PLM. Config looks like:

```
{
  "platform": "InsteonLocal",
  "name": "Insteon",
  "host": "/dev/ttyUSB0",
  "refresh": "300",
  "server_port": "3000",
  "keepAlive": "3600",
  "devices": [
    {
      "deviceID": "3da13f",
      "name": "Bedroom lights",
      "deviceType": "dimmer",
      "dimmable" : "yes"
    },
    {
      "deviceID": "3da075",
      "name": "Nursey lights",
      "deviceType": "dimmer",
      "dimmable" : "yes"
    }
  ]
}
```

Moving the dimmer up & down is temperamental, likely because of the debouncing that needs to be done. Often the light ends up at 100% brightness after moving the dimmer up & down a bit.

Here's some noisy log file output:

```
homebridge_1  | 2019-02-15T07:09:47.178Z home-controller:insteon canceling - match: /^....35571C.*/i
homebridge_1  | [2/15/2019, 7:09:47 AM] [Insteon] Setting level of Marquee 'G' sign to undefined%
homebridge_1  | 2019-02-15T07:09:47.179Z home-controller:insteon Send: {"command":{"cmd1":"19","cmd2":"00","type":"62","raw":"026235571C0F1900","id":"35571C"},"deferred":{"promise":{}},"timeout":5000}
homebridge_1  | 2019-02-15T07:09:47.180Z home-controller:insteon canceling - match: /^....35571C.*/i
homebridge_1  | [2/15/2019, 7:09:47 AM] [Insteon] Setting level of Marquee 'G' sign to 100%
homebridge_1  | 2019-02-15T07:09:47.181Z home-controller:insteon Already running a command - queuing command: {"cmd1":"21","cmd2":"FF","type":"62","raw":"026235571C0F21FF","id":"35571C"}
homebridge_1  | 2019-02-15T07:09:47.189Z serialport:main _write 8 bytes of data
homebridge_1  | 2019-02-15T07:09:47.189Z serialport:bindings write 8 bytes
homebridge_1  | 2019-02-15T07:09:47.189Z serialport:unixWrite Starting write 8 bytes offset 0 bytesToWrite 8
homebridge_1  | 2019-02-15T07:09:47.190Z serialport:unixWrite write returned null 8
homebridge_1  | 2019-02-15T07:09:47.190Z serialport:unixWrite wrote 8 bytes
homebridge_1  | 2019-02-15T07:09:47.190Z serialport:unixWrite Finished writing 8 bytes
homebridge_1  | 2019-02-15T07:09:47.190Z serialport:main binding.write write finished
homebridge_1  | 2019-02-15T07:09:47.196Z serialport:poller received "readable"
homebridge_1  | 2019-02-15T07:09:47.196Z serialport:bindings read
homebridge_1  | 2019-02-15T07:09:47.196Z serialport:unixRead Starting read
homebridge_1  | 2019-02-15T07:09:47.196Z serialport:unixRead Finished read 5 bytes
homebridge_1  | 2019-02-15T07:09:47.196Z serialport:main binding.read finished
homebridge_1  | 2019-02-15T07:09:47.196Z home-controller:insteon Rcvd: 026235571c
homebridge_1  | 2019-02-15T07:09:47.196Z home-controller:insteon checkStatus - buffer: 026235571c
homebridge_1  | 2019-02-15T07:09:47.196Z home-controller:insteon checkStatus - status: {"command":{"cmd1":"19","cmd2":"00","type":"62","raw":"026235571C0F1900","id":"35571C"},"deferred":{"promise":{}},"timeout":5000}
homebridge_1  | 2019-02-15T07:09:47.196Z home-controller:insteon parsing - insufficient data to continue - 10
```

Interesting thing is the `undefined%` value that is logged.

**Comment by kuestess** [comment]
@bradgessler I just pushed a fix that I think will address this issue - give it a try and see if that takes care of it.

**Comment by bradgessler** [comment]
This helped when controlling one device. So much better! Thank you.

When I try to get it working with multiple lights (two in this case to keep things simple) for a scene, either one of the lights either doesn’t turn on or it turns on 100%. Is the debouncing maybe scoped globally and not per device or Homekit request?
